src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs

```diff
@@ -33,9 +35,36 @@ public abstract class FormulaType
         public static FormulaType OptionSetValue { get; } = new OptionSetValueType();
 
         // chained by derived type 
-        internal FormulaType(DType type)
+        internal FormulaType(DType type, Dictionary<string, string> displayNameSet = null)
         {
-            _type = type;
+            if (displayNameSet == null)
+                _type = type;
+            else 
+            {
+                ValidateDisplayNames(displayNameSet);
+                _type = DType.AttachDisplayNameProvider(type, new BidirectionalDictionary<string, string>(displayNameSet));
+            }
+
+        }
+
+        private void ValidateDisplayNames(Dictionary<string, string> displayNameSet)
+        {
+            HashSet<string> displayNames = new();
+            foreach (var kvp in displayNameSet)
+            {
+                // Display names can't collide with other display names
+                if (displayNames.Contains(kvp.Value))
+                {
+                    throw new DisplayNameCollisionException(kvp.Value);
+                }
+                // Display names can't collide with logical names other than their own
+                if (kvp.Key != kvp.Value && displayNameSet.ContainsKey(kvp.Value))
```

Comment:
>!= [](http://example.com/codeflow?start=28&length=2)

What are the string casing rules?

Reply:
Added comment calling it out, we're case-sensitive